### PR TITLE
Week 6

### PR DESCRIPTION
**Pull request by aishna2502**
Please Review! Thank you in advance!

### CODE REVIEW

File: TreeMatching.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	
+	@Test
+	public void FalseOneNodeTest(){
+		TreeNode root = new TreeNode(1);
+		TreeNode root1 = new TreeNode(7);
+		assertEquals(false, matchTree(root1, root));
+	}
+	
+	@Test
+	public void FalseSubTreeTest(){
```
Comment (mooreaarond) [inline TreeMatching.java:126]: Looks like you got the names of these two names tests mixed up :)

File: TreeMatching.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+		return matchTree(root1.left, root2) || matchTree(root1.right, root2);
+	}
+	
+//	TEST CASES
+	
+	@Test
+	public void NullTest(){
+		TreeNode root1 = null;
+		TreeNode root2 = null;	
+		assertEquals(true, matchTree(root1, root2));
```
Comment (mooreaarond) [inline TreeMatching.java:76]: Another test case might be if only one the the trees is null instead of both

Comment (mooreaarond) [reply TreeMatching.java:76]: A question I might ask the interviewer: if root 1 is nonempty and root2 is NULL, is the answer true?  I think this might be true in the same way the empty set is a subset of any set.  I don't think your solution covers this case.

File: Palindrome.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		start = prev;
+		return start;
+	}
+	
+	public static boolean palindrome(Node start){
+		if(start==null){
+			return true;
+		}
+		if(start.next==null){
+			return false;
```
Comment (mooreaarond) [inline Palindrome.java:48]: One element list should be a palindrome, right?
"1" is the same forward and backwards

File: Palindrome.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+		}
+		Node last = start;
+		Node temp = start;
+		int n = (length/2);
```
Comment (mooreaarond) [inline Palindrome.java:52]: For this problem, you can (and should) solve it without being given this information.  You should also avoid using static variables unless absolutely necessary as they can be changed at any time and make programs difficult to reason about.

https://stackoverflow.com/questions/7026507/why-are-static-variables-considered-evil

File: Palindrome.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+import static org.testng.Assert.assertEquals;
+
+//import java.util.Scanner;
+
+import org.junit.Test;
+
+public class Palindrome {
+	static int length = 0;
+	
+	public static boolean is_palindrome(Node first, Node second){
```
Comment (mooreaarond) [inline Palindrome.java:12]: is_palindrome and palindrome are basically the same name and it would be hard to tell which to use in what situation.  It looks like this function is checking whether the two lists are equal, so maybe something like AreEqual or something like that

File: Palindrome.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+		}
+		if(start.next==null){
+			return false;
+		}
+		Node last = start;
+		Node temp = start;
+		int n = (length/2);
+		int l = 0;
+		while(l<n){
+			l++;
```
Comment (mooreaarond) [inline Palindrome.java:55]: if you know how much you are iterating (0<=i<n), you should be using a for loop here.

File: Palindrome.java (hunk 4 above)
Comment (mooreaarond) [reply Palindrome.java:52]: https://stackoverflow.com/questions/16446355/how-to-find-a-middle-element-in-a-link-list-without-traversing-the-entire-list

File: Palindrome.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	
+
+	@Test
+	public void emptyPali(){
+		Node n = null;
+		assertEquals(true, palindrome(n));
+	}
+	
+	@Test
+	public void oddPaliRight(){
```
Comment (mooreaarond) [inline Palindrome.java:77]: Good test cases!

File: Palindrome.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+		n2.next = n3;
+		n3.next = n4;
+		assertEquals(false, palindrome(n1));
+	}
+	
+	@Test
+	public void onePaliWrong(){
+		Node n1 = new Node(1);
+		length = 1;
+		assertEquals(false, palindrome(n1));
```
Comment (mooreaarond) [inline Palindrome.java:128]: As mentioned above, this should be true!  "1" is the same forwards and backwards

**Review by mooreaarond — COMMENTED** [review]
Your algorithms, code and test cases look good for both of these problems!  I've left you a few comments, but good work this week
